### ISSUE TITLE
Issues using a custom pretrained Huggingface model (loading safetensors)

### ISSUE THREAD

**Issue by hunklinger** [Acellera/acegen-open#51]
First, thanks for the amazing repository! I wanted to load a pretrained model from Huggingface, which typically creates a folder with the `config.json` and the .bin file containing the weights inside. In your workflow I see you only use checkpoints in the .ckpt file format. I tried to modify the snippet in which the weights are loaded from `ckpt = torch.load(ckpt_path, map_location=device, weights_only=True)` to `ckpt = safetensors.torch.load_file(ckpt_path)` to adjust to the file format change. Its not crashing but its not aligning the content correctly as I am getting a warning for every loaded tensor. As you can see its trying to load the h11.mlp.c_fc.weight as a h3.ln_2.bias?! The shapes therefore don't match.
```
miniconda3/envs/acegen/lib/python3.10/site-packages/acegen/models/utils.py:27: UserWarning: The shape of source key transformer.h.11.mlp.c_fc.weight (torch.Size([1280, 5120])) and target key module.0.module.0.module.feature_extractor.transformer.h.3.ln_2.bias (torch.Size([1280])) do not match.
  warnings.warn(
miniconda3/envs/acegen/lib/python3.10/site-packages/acegen/models/utils.py:27: UserWarning: The shape of source key transformer.h.11.mlp.c_proj.bias (torch.Size([1280])) and target key module.0.module.0.module.feature_extractor.transformer.h.3.mlp.c_fc.weight (torch.Size([1280, 5120])) do not match.
  warnings.warn(
miniconda3/envs/acegen/lib/python3.10/site-packages/acegen/models/utils.py:27: UserWarning: The shape of source key transformer.h.11.mlp.c_proj.weight (torch.Size([5120, 1280])) and target key module.0.module.0.module.feature_extractor.transformer.h.3.mlp.c_fc.bias (torch.Size([5120])) do not match.
  warnings.warn(
miniconda3/envs/acegen/lib/python3.10/site-packages/acegen/models/utils.py:27: UserWarning: The shape of source key transformer.h.12.attn.c_attn.bias (torch.Size([3840])) and target key module.0.module.0.module.feature_extractor.transformer.h.3.mlp.c_proj.weight (torch.Size([5120, 1280])) do not match.
  warnings.warn(
miniconda3/envs/acegen/lib/python3.10/site-packages/acegen/models/utils.py:27: UserWarning: The shape of source key transformer.h.12.attn.c_attn.weight (torch.Size([1280, 3840])) and target key module.0.module.0.module.feature_extractor.transformer.h.3.mlp.c_proj.bias (torch.Size([1280])) do not match.
...
```
I did add my custom model according to the option 1 described in your tutorials and generated an actor and critic according to the provided config file from the models Huggingface page. I also tried converting the safetensor to a checkpoint file before using it but that did not work either. I was wondering if you had any idea where the issue is coming from and how I can fix it. Is maybe the added linear layer during the actor and critic creation messing with the checkpoint loading later on?

Another thought: Would it be possible to somehow use the standard `from_pretrained()` method from Huggingface? I guess that would need to happen inside the `models/custom_model.py` script before the `TensorDictModule` wrapping.

**Comment by albertbou92** [comment]
Hi!

I haven’t previously used pretrained weights from Huggingface, but I think it makes sense to improve their integration into the repo. It is probably a common use case.

I’ve encountered similar weight loading issues before. The fact that it doesn’t crash, as you mentioned, suggests that the only problem is the mismatch in architectures. Your models should be correct in all other aspects.

The cause is generally a minor problem, but difficult to identify. Yes, could be that adding and extra layers changes the order of the checkpoint layers. But you could modify the checkpoint to match the order if needed. What I would try is the following:

Here is what I would do:
- Make sure the huggingface version used to save he checkpoint is the same you are using the load it (huggingface can change the architectures a bit between versions)
- I would then start with an algorithm that does not need a critic, like Reinforce. You could even load the weights first within the class. It is not very flexible, but you can later modify it.
- When moving up to an algorithms that needs a critic, like ppo, you can first try using separate networks for the actor and the critic.
- Finally, when using a shared actor-critic architecture simply go over the weights of both you model and the checkpoint and make sure order is the same. If the order is different, you could manually change the checkpoint.

```
for name, param in actor.named_parameters():
    print(f"Name: {name}, Shape: {param.shape}")
```

Let me know if I can help you further.

**Comment by hunklinger** [comment]
Thanks so much for the quick response! I will try your suggestions and get back to you soon.

**Comment by hunklinger** [comment]
I have tried a couple of things now and its not producing any loading specific error anymore but from the decoded output I am getting I assume the model is still not correctly implemented.
Did I understand you correctly that when you were saying "load the weights first within the class using from_pretrained()" that I won't let the system override them within the reinforce.py script anymore and basically modify this script like this to skip that part: 
```
if cfg.model != 'custom_model':
        ckpt = torch.load(ckpt_path, map_location=device, weights_only=True)
        actor_inference.load_state_dict(
            adapt_state_dict(ckpt, actor_inference.state_dict())
        )
        actor_training.load_state_dict(adapt_state_dict(ckpt, actor_training.state_dict()))
```
Do you have any idea why the model is still not behaving like it does when I load it in my other scripts. It should not be modified in the TensorDict wrapping right? The main change I did was going from this line `self.feature_extractor = transformers.GPT2Model(config)` to the one including the from_pretrained(<Hub-handle>) in the __ init __ function of the custom model class. And in the forward() function the .last_hidden_state attribute was not found anymore and I made the workaround using `out = out.hidden_states[-1] ` which I hope is correct.
Also, I think it was a very good comment to start with a more simple algorithm than PPO and work your way up. Thanks.

**Comment by albertbou92** [comment]
Yes, I think that replacing the `self.feature_extractor = transformers.GPT2Model(config)` for a `from_pretrained()` method is correct. However, note that after the feature extractor there is a final layer that predicts the logits of the tokens, do you have weights to load for this layer as well? What was your model pretrained on? Our scripts pretrain to do next-token prediction, and this learn this last layer too

**Comment by hunklinger** [comment]
That totally makes sense. I actually noticed that the pretrained model I am using already has a linear layer at the end so I deleted the lines you refer to for my custom model as I don't need it and would have to train the weights again. After all, it seems like now the model is generating how I would suspect. Thank you so much for your (very quick) help!
